man/sd_bus_service_reconnect.c

```diff
@@ -0,0 +1,204 @@
+
+  object *o = userdata;
+
+  if (strcmp(property, "Example") == 0)
+    return sd_bus_message_append(reply, "s", o->example);
+
+  return sd_bus_error_setf(error,
+                           SD_BUS_ERROR_UNKNOWN_PROPERTY,
+                           "Unknown property '%s'",
+                           property);
+}
+
+/* https://www.freedesktop.org/software/systemd/man/sd_bus_add_object.html */
+static const sd_bus_vtable vtable[] = {
+  SD_BUS_VTABLE_START(0),
+  SD_BUS_PROPERTY(
+    "Example", "s",
+    property_get,
+    0,
+    SD_BUS_VTABLE_PROPERTY_CONST),
+  SD_BUS_VTABLE_END
+};
+
+static int setup(object *o);
+
+static int on_disconnect(sd_bus_message *message, void *userdata, sd_bus_error *ret_error) {
+  check(setup((object *)userdata));
+  return 0;
+}
+
+static int setup(object *o) {
+  /* If we are reconnecting, then the bus object needs to be closed, detached from
+   * the event loop and recreated.
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_detach_event.html
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_close_unref.html
+   */
+  if (*o->bus) {
+    check(sd_bus_detach_event(*o->bus));
+    *o->bus = sd_bus_close_unref(*o->bus);
+  }
+
+  /* Set up a new bus object for the system bus, configure it to wait for D-Bus to be available
+   * instead of failing if it is not, and start it.
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_new.html
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_set_address.html
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_set_bus_client.html
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_negotiate_creds.html
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_set_watch_bind.html
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_set_connected_signal.html
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_start.html
+   */
+  check(sd_bus_new(o->bus));
+  check(sd_bus_set_address(*o->bus, "unix:path=/run/dbus/system_bus_socket"));
+  check(sd_bus_set_bus_client(*o->bus, 1));
+  check(sd_bus_negotiate_creds(*o->bus, 1, SD_BUS_CREDS_UID|SD_BUS_CREDS_EUID|SD_BUS_CREDS_EFFECTIVE_CAPS));
+  check(sd_bus_set_watch_bind(*o->bus, 1));
+  check(sd_bus_set_connected_signal(*o->bus, 1));
+  check(sd_bus_start(*o->bus));
+
+  /* Publish an interface on the bus, specifying our well-known object access
+   * path and public interface name.
+   * https://www.freedesktop.org/software/systemd/man/sd_bus_add_object.html
+   * https://dbus.freedesktop.org/doc/dbus-tutorial.html
+   */
+  check(sd_bus_add_object_vtable(*o->bus,
+                                 NULL,
+                                 "/org/freedesktop/ReconnectExample",
+                                 "org.freedesktop.ReconnectExample",
+                                 vtable,
+                                 o));
+  /* By default the service is assigned an ephemeral name. Also add a fixed
```

Comment:
*only* assigned.

services always have the ephemeral name (which actually is called "unique name" in D-Bus speak), they can acquire zero, one or more well-known names in addition.

and yes, "fixed names" are actually called "well-known names" in D-Bus speak